src/storage/quota/QuotaStrategy.ts

```diff
@@ -0,0 +1,46 @@
+import type { PassThrough, Readable } from 'stream';
+import type { RepresentationMetadata } from '../../http/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../../http/representation/ResourceIdentifier';
+import type { Guarded } from '../../util/GuardedStream';
+import type { Size } from '../size-reporter/Size';
+
+/**
+ * A QuotaStrategy is used when we want to set a limit to the amount of data that can be
+ * stored on the server.
+ * This can range from a limit for the whole server to a limit on a per pod basis.
+ * The way the size of a resource is calculated is implemented by the implementing classes.
+ * This can be bytes, quads, file count, ...
+ */
+export interface QuotaStrategy {
+  /**
+   * Get the available space given a resource's identifier.
+   * If the given resource already exists it will deduct the already taken up
+   * space by that resource since it is going to be overwritten and thus counts
+   * as available space.
+   *
+   * @param identifier - the identifier of the resource of which you want the available space
+   * @returns the available space and the unit of the space as a Size object
+   */
+  getAvailableSpace: (identifier: ResourceIdentifier) => Promise<Size>;
+
+  /**
+   * Get an estimated size of the resource
+   *
+   * @param metadata - the metadata that might include the size
+   * @returns a Size object containing the estimated size and unit of the resource
+   */
+  estimateSize: (metadata: RepresentationMetadata) => Promise<Size>;
+
+  /**
+   * Get a Passthrough stream that will keep track of the available space.
+   * If the quota is exceeded the stream will be destroyed.
+   * Like other Passthrough instances this will simply pass on the chunks, when the quota isn't exceeded.
+   *
+   * @param identifier - the identifier of the resource in question
+   * @param data - the Readable stream that belongs to the identifier
+   * @param metadata - the RepresentationMetadata that belongs to the identifier
+   * @returns a Passthrough instance that errors when quota is exceeded
+   */
+  trackAvailableSpace: (identifier: ResourceIdentifier, data: Guarded<Readable>, metadata: RepresentationMetadata) =>
```

Comment:
The function returns a Passthrough that keeps track of the available space and the used space by the stream so I don't think `trackStreamSize` fits.
The only other thing that comes to mind is `ensureAvailableSpace` since it ensures there is space more then it tracks the actual available space, but that might be just as bad as `trackAvailableSpace` tbh.

@woutermont can you think of a more fitting name ?

Reply:
What specifically do you think is wrong with the current name, @joachimvh? The function returns a stream that tracks the available space for an identifier. Alternative ideas might be `getAvailableSpaceCheck` or `createSpaceGuard`, `makeSizeLimitedStream` ...

Reply:
What this function does is:
 * Create a new stream that errors once it has seen a certain amount of data pass by.
 * Pipe the input stream into this new stream (and return the new stream).

My with it was that I feel like the name does not really convey that a new stream is being created by this function call. But this is a very minor remark since the behaviour is described in the comment above. I do like the suggestion of having the name start with `create`, but `createSpaceGuard` makes me think more of science fiction than a quota strategy 😅. Perhaps `createQuotaGuard`? But if you feel any of the other names convey the behaviour better feel free to take those, I'm ok with anything that is related to this (even `trackAvailableSpace`), just wanted to convey the feeling I had with this name.

Reply:
I think `createQuotaGuard` if fine!

_And so, after creating a Space Guard for the Great Pod, he boarded the CSS, and once again set sail into the Solidverse ..._